src/update-feeds-loops/get-updatable-feeds.ts

```diff
@@ -54,15 +54,36 @@ export const getUpdatableFeeds = (
           isUpdatable ? offChainValue! : onChainValue
         );
 
-        const newBeaconSetValue = calculateMedian(beaconValues.map(({ value }) => value));
-        const newBeaconSetTimestamp = calculateMedian(beaconValues.map(({ timestamp }) => timestamp));
+        // Works for both beacon sets and single beacon feed.
+        const newDataFeedValue = calculateMedian(beaconValues.map(({ value }) => value));
+        const newDataFeedTimestamp = calculateMedian(beaconValues.map(({ timestamp }) => timestamp));
 
-        const { decodedUpdateParameters, dataFeedValue, dataFeedTimestamp, decodedDapiName, dataFeedId } = dataFeedInfo;
+        const {
+          decodedUpdateParameters: { deviationThresholdInPercentage, deviationReference, heartbeatInterval },
+          dataFeedValue,
+          dataFeedTimestamp,
+          decodedDapiName,
+          dataFeedId,
+        } = dataFeedInfo;
         const adjustedDeviationThresholdCoefficient = multiplyBigNumber(
-          decodedUpdateParameters.deviationThresholdInPercentage,
+          deviationThresholdInPercentage,
           deviationThresholdCoefficient
         );
 
+        // We need to make sure that the update transaction actually changes the on-chain value. There are two cases:
+        // 1. Data feed is a beacon - We need to make sure the off-chain data updates the feed. This requires timestamp
+        //    to change. See:
+        //    https://github.com/api3dao/airnode-protocol-v1/blob/65a77cdc23dc5434e143357a506327b9f0ccb7ef/contracts/api3-server-v1/DataFeedServer.sol#L120
+        // 2. Data feed is a beacon set - We need make sure that the beacon set will change. The contract requires the
+        //    beacon set value or timestamp to change. See:
+        //    https://github.com/api3dao/airnode-protocol-v1/blob/65a77cdc23dc5434e143357a506327b9f0ccb7ef/contracts/api3-server-v1/DataFeedServer.sol#L54
+        const isBeaconSet = beaconValues.length > 1;
+        // Note, that the beacon set value/timestamp is computed as median, so single beacon update may not result in a beacon set update.
+        if (isBeaconSet && newDataFeedValue === dataFeedValue && newDataFeedTimestamp === dataFeedTimestamp) {
+          return false;
+        }
+        if (!isBeaconSet && newDataFeedTimestamp === dataFeedTimestamp) return false;
+
```

Comment:
I'm just nit-picking 😄. Because this snippet is directly related to `isDataFeedUpdatable`, it'd be better to place it into that function. Maybe that function can accept extra parameter `isBeaconSet`. 

Addition to that, for this condition `!isBeaconSet && newDataFeedTimestamp === dataFeedTimestamp`, it's hard for one to understand that this equals to `newDataFeedTimestamp <= dataFeedTimestamp` (coming from [here](https://github.com/api3dao/airseeker-v2/blob/6bd46b286769d89b595e7d22aa1c8344c04522a7/src/update-feeds-loops/get-updatable-feeds.ts#L41)). So maybe we can just write this condition explicitly again like `!isBeaconSet && newDataFeedTimestamp <= dataFeedTimestamp`.

Reply:
> Because this snippet is directly related to isDataFeedUpdatable, it'd be better to place it into that function. Maybe that function can accept extra parameter isBeaconSet.

I was thinking about that and the problem I have with that was that it would stop mimicking the contract implementation of deviation logic. I think in the source of truth for the update condition is `getUpdatableFeeds` and not the deviation checks because even if we move the timestamp checks there, making sure we only update the beacons for which we have newer data is as important (and that happens outside of deviation checks).

> Addition to that, for this condition !isBeaconSet && newDataFeedTimestamp === dataFeedTimestamp, it's hard for one to understand that this equals to newDataFeedTimestamp <= dataFeedTimestamp (coming from [here](https://github.com/api3dao/airseeker-v2/blob/6bd46b286769d89b595e7d22aa1c8344c04522a7/src/update-feeds-loops/get-updatable-feeds.ts#L41)). So maybe we can just write this condition explicitly again like !isBeaconSet && newDataFeedTimestamp <= dataFeedTimestamp.

What I dislike about that is that it makes things more confusing once you understand what's happening - for example, the timestamp check I originally removed seemed extraneous because it was checking whether off-chain timestamp is smaller than on-chain which could never occur.

Reply:
I tried to clarify this in the comments and updated the deviation checks function to stress it more.